File: packages/coast_audio_miniaudio/mabridge/src/mab_low_shelf_filter.c
#define MINIAUDIO_IMPLEMENTATION
#include "miniaudio.h"

#include "mab_types.h"
#include "mab_low_shelf_filter.h"

mab_low_shelf_filter_config mab_low_shelf_filter_config_init(mab_format format, u_int32_t sampleRate, u_int32_t channels, double gainDb, double shelfSlope, double frequency)
{
  mab_low_shelf_filter_config config = {
      .format = format,
      .sampleRate = sampleRate,
      .channels = channels,
      .gainDb = gainDb,
      .shelfSlope = shelfSlope,
      .frequency = frequency,
  };
  return config;
}

mab_result mab_low_shelf_filter_init(mab_low_shelf_filter* pLSF, mab_low_shelf_filter_config config)
{
  ma_loshelf2_config maConfig = ma_loshelf2_config_init(mab_cast(ma_format, config.format), config.channels, config.sampleRate, config.gainDb, config.shelfSlope, config.frequency);
  ma_loshelf2* pData = MAB_MALLOC(sizeof(ma_loshelf2));
  ma_result result = ma_loshelf2_init(&maConfig, NULL, pData);
  if (result != MA_SUCCESS) {
    MAB_FREE(pData);
    return result;
  }

  pLSF->pData = pData;
  return MA_SUCCESS;
}

mab_result mab_low_shelf_filter_process(mab_low_shelf_filter* pLSF, void* pFramesOut, const void* pFramesIn, u_int64_t frameCount)
{
  ma_loshelf2 *pData = (ma_loshelf2*)pLSF->pData;
  return ma_loshelf2_process_pcm_frames(pData, pFramesOut, pFramesIn, frameCount);
}

mab_result mab_low_shelf_filter_reinit(mab_low_shelf_filter* pLSF, mab_low_shelf_filter_config config)
{
  ma_loshelf2 *pData = (ma_loshelf2*)pLSF->pData;
  ma_loshelf2_config maConfig = ma_loshelf2_config_init(mab_cast(ma_format, config.format), config.channels, config.sampleRate, config.gainDb, config.shelfSlope, config.frequency);
  return ma_loshelf2_reinit(&maConfig, pData);
}

u_int32_t mab_low_shelf_filter_get_latency(mab_low_shelf_filter* pLSF)
{
  ma_loshelf2 *pData = (ma_loshelf2*)pLSF->pData;
  return ma_loshelf2_get_latency(pData);
}

void mab_low_shelf_filter_uninit(mab_low_shelf_filter* pLSF)
{
  ma_loshelf2 *pData = (ma_loshelf2*)pLSF->pData;
  ma_loshelf2_uninit(pData, NULL);
  MAB_FREE(pData);
  pLSF->pData = NULL;
}

File: packages/coast_audio_miniaudio/lib/src/ma_bridge/filters/mab_low_shelf_filter.dart
import 'dart:ffi';

import 'package:coast_audio/coast_audio.dart';
import 'package:coast_audio_miniaudio/generated/ma_bridge_bindings.dart';
import 'package:coast_audio_miniaudio/src/ma_bridge/filters/mab_filter_base.dart';
import 'package:coast_audio_miniaudio/src/ma_extension.dart';

class MabLowShelfFilter extends MabFilterBase {
  MabLowShelfFilter({
    required this.format,
    required double gainDb,
    required double shelfSlope,
    required double frequency,
    super.memory,
  })  : _gainDb = gainDb,
        _shelfSlope = shelfSlope,
        _frequency = frequency {
    final config = library.mab_low_shelf_filter_config_init(format.sampleFormat.mabFormat.value, format.sampleRate, format.channels, gainDb, shelfSlope, frequency);
    library.mab_low_shelf_filter_init(_pLSF, config).throwMaResultIfNeeded();
    _updateLatency();
  }

  final AudioFormat format;

  double _gainDb;
  double get gainDb => _gainDb;

  double _shelfSlope;
  double get shelfSlope => _shelfSlope;

  double _frequency;
  double get frequency => _frequency;

  late AudioTime _latency;
  AudioTime get latency => _latency;

  late final _pLSF = allocate<mab_low_shelf_filter>(sizeOf<mab_low_shelf_filter>());

  @override
  void process(AudioBuffer bufferOut, AudioBuffer bufferIn) {
    assert(bufferOut.sizeInFrames >= bufferIn.sizeInFrames);
    library.mab_low_shelf_filter_process(_pLSF, bufferOut.pBuffer.cast(), bufferIn.pBuffer.cast(), bufferIn.sizeInFrames).throwMaResultIfNeeded();
  }

  void reinit({double? gainDb, double? shelfSlope, double? frequency}) {
    final config = library.mab_low_shelf_filter_config_init(
      format.sampleFormat.mabFormat.value,
      format.sampleRate,
      format.channels,
      gainDb ?? this.gainDb,
      shelfSlope ?? this.shelfSlope,
      frequency ?? this.frequency,
    );
    library.mab_low_shelf_filter_reinit(_pLSF, config).throwMaResultIfNeeded();
    _gainDb = gainDb ?? this.gainDb;
    _shelfSlope = shelfSlope ?? this.shelfSlope;
    _frequency = frequency ?? this.frequency;
    _updateLatency();
  }

  void _updateLatency() {
    final frameCount = library.mab_low_shelf_filter_get_latency(_pLSF);
    _latency = AudioTime.fromFrames(frames: frameCount, format: format);
  }

  @override
  void uninit() {
    library.mab_low_shelf_filter_uninit(_pLSF);
  }
}
